Index: marketing-tools/src/main/resources/application.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>spring:\r\n  application:\r\n    name: marketing-tools\r\nserver:\r\n  port: 19777\r\notherConfig:\r\n  alarm:\r\n    secretKey: a9c57e4138807a9e18b439cd9dab275b\r\n    appName: marketing\r\n    outsideSecretKey: 71e63ead0af0a07ae4191eb14b75a541\r\n    outsideAppName: marketing-out\r\n  userCenter:\r\n    appName: loan-strategy\r\n    appSecretKey: 404be42e9ad8ce59af79f17eb51594cf\r\n  warning:\r\n    pageSize: 5000\r\n    apiUrl: http://compass-api.100credit.com/api/loan-warning-api/loanStrategy/v2/query\r\n    path: /opt/data/inloan/download/marketing/\r\n    sftpHost: 10.100.123.65\r\n    sftpPort: 9999\r\n    sftpUser: BRmarketing_pre\r\n    sftpPwd: 72aXd79ITcMeuyPAlDYy@_2021.06.07\r\n  encodeMapping:\r\n    appName: loan-warning\r\n    appSecretKey: a938d0e974746651a11ddcdbcea31404
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/marketing-tools/src/main/resources/application.yml b/marketing-tools/src/main/resources/application.yml
--- a/marketing-tools/src/main/resources/application.yml	
+++ b/marketing-tools/src/main/resources/application.yml	
@@ -22,4 +22,4 @@
     sftpPwd: 72aXd79ITcMeuyPAlDYy@_2021.06.07
   encodeMapping:
     appName: loan-warning
-    appSecretKey: a938d0e974746651a11ddcdbcea31404
\ No newline at end of file
+    appSecretKey: f15a1421ac95a284683688af14688b72
\ No newline at end of file
